Use React Query for data fetching and error handling

ningxie1991 opened the pull request:
For issue #56.

This is to use React Query to catch errors in data fetching and display the error on the UI without crashing the app. 

@chaoran-chen @TKGZ 
As discussed in issue #56, Error Boundary is not applicable for functional components using the `useEffect` hook. And @TKGZ suggested looking into [React Query](https://react-query.tanstack.com/). I checked some blog posts and the documentations and examples. Basically we can replace the whole saving response data in a local state to using `useQuery` which also catches errors and returns the error object in the result. 

Error handling has been implemented for the following api requests:

`getCaseCounts`

![image](https://user-images.githubusercontent.com/3387698/131342897-1eab62f7-1953-4975-8d3b-d4730fbe2750.png)

`getSamples` 

![image](https://user-images.githubusercontent.com/3387698/131343418-d2d83419-b351-4b5a-9ef9-4193883e2d43.png)

`getCaseCounts` + `getSequenceCounts`

![image](https://user-images.githubusercontent.com/3387698/131340776-9e3e9234-ccbf-472c-aa4a-25a00a38a605.png)

`getPangolinLineages`

![image](https://user-images.githubusercontent.com/3387698/131342748-15eb5d2a-bf6d-4cf2-b2c6-e3b97e7feb9b.png)

getPangolinLineageAliases

https://github.com/cevo-public/cov-spectrum-website/blob/5f99d478ec7339a3f565217ac5ff8e3e6df0b507/src/services/PangolinLineageAliasResolverService.ts#L11-L21

ningxie1991 committed:
Try out React Query for data fetching and error handling

ningxie1991 committed:
Fix formatting

ningxie1991 committed:
Update package-lock.json

ningxie1991 committed:
Update SwitzerlandEstimatedCasesDivisionModal.tsx

ningxie1991 commented on src/components/SwitzerlandEstimatedCasesDivisionModal.tsx:
```diff
@@ -87,26 +88,23 @@ export const SwitzerlandEstimatedCasesDivisionModal = ({
   handleClose,
 }: Props) => {
-  const [caseCounts, setCaseCounts] = useState<CaseCountEntry[] | undefined>();
+  //const [caseCounts, setCaseCounts] = useState<CaseCountEntry[] | undefined>();
```
We can avoid saving the response data in a local state (see [blog post](https://blog.bitsrc.io/how-to-start-using-react-query-4869e3d5680d) )

ningxie1991 commented on src/components/SwitzerlandEstimatedCasesDivisionModal.tsx:
```diff
@@ -87,26 +88,23 @@ export const SwitzerlandEstimatedCasesDivisionModal = ({
-
   const { dateFrom, dateTo, country } = variantSampleSet.sampleSelector;
-  useEffect(() => {
-    let isSubscribed = true;
```
I am not sure why we initialize `isSubscribed` to true here but the variable is to be checked before `setCaseCounts` at the deleted line 100, which will always be true.

ningxie1991 commented on src/components/SwitzerlandEstimatedCasesDivisionModal.tsx:
```diff
@@ -87,26 +88,23 @@ export const SwitzerlandEstimatedCasesDivisionModal = ({
-      controller.abort();
-    };
-  }, [dateFrom, dateTo, country]);
+  const { isLoading, isSuccess, error, isError, data: caseCounts, refetch } = useQuery<
```
We can use these properties returned from `useQuery` to check if the request `isLoading`, `isSuccess` or `isError`, which catches any error in the request.

ningxie1991 commented on src/components/SwitzerlandEstimatedCasesDivisionModal.tsx:
```diff
@@ -151,8 +149,18 @@ export const SwitzerlandEstimatedCasesDivisionModal = ({
-      {cantonData && regionData && (
+
+      {isLoading && <Loader />}
+      {isError && error && (
```
We can simply use `isError` and `error` to display the error messages when there's an error in the fetching of data.

ningxie1991 commented on src/pages/FocusPage.tsx:
```diff
@@ -315,12 +318,14 @@ export const FocusPage = ({
-          show={showEstimatedCasesDivGrid}
-          handleClose={() => setShowEstimatedCasesDivGrid(false)}
-        />
+        <QueryClientProvider client={queryClient}>
```
We have to enclose the functional component with this `<QueryClientProvider>` tag which I think can be placed at a even higher level like on the whole page component or App component

ningxie1991 committed:
fix formatting

ningxie1991 commented on src/components/SwitzerlandEstimatedCasesDivisionModal.tsx:
```diff
@@ -87,26 +88,23 @@ export const SwitzerlandEstimatedCasesDivisionModal = ({
+  >('caseCounts', fetchCaseCounts);
+
+  useEffect(() => {
+    refetch();
```
We can call the `refetch` function which is returned from the `useQuery` in the useEffect hook whenever `dateFrom`, `dateTo` and `country` is updated.

ningxie1991 commented on src/components/SwitzerlandEstimatedCasesDivisionModal.tsx:
```diff
@@ -87,26 +88,23 @@ export const SwitzerlandEstimatedCasesDivisionModal = ({
+  };
 
-    return () => {
-      isSubscribed = false;
```
@chaoran-chen @TKGZ What is the purpose of `isSubscribed` ?

spacephoton replied:
This prevents duplicate fetches, for example if a new fetch is called before the previous one finishes.

chaoran-chen replied:
@ningxie1991, we have some queries that can take quite some time. Do you know what will now happen if the user navigates to another page before the data for the first one was fully loaded? I.e. if the first query is finished after the second query?

ningxie1991 replied:
I updated the code so that react query behaves the same. It will cancel request if we close or click out of the modal. And `refetch` data in `useEffect` if `isFetching` = false 

![image](https://user-images.githubusercontent.com/3387698/130066887-9d50fac9-4217-4a95-8fd3-face5e4903a9.png)

ningxie1991 committed:
Update api.ts

ningxie1991 committed:
Merge with changes from develop branch

spacephoton reviewed (approved):
Seems good!

ningxie1991 committed:
Merge branch 'develop' into ui-error-handling

ningxie1991 committed:
Update SwitzerlandEstimatedCasesDivisionModal.tsx

Added code to cancel request when the modal is closed or user navigates away from the modal

ningxie1991 committed:
Update SwitzerlandEstimatedCasesDivisionModal.tsx

Remove wrong dependencies

ningxie1991 commented on src/components/SwitzerlandEstimatedCasesDivisionModal.tsx:
```diff
@@ -88,23 +92,25 @@ export const SwitzerlandEstimatedCasesDivisionModal = ({
+    const controller = new AbortController();
+    const signal = controller.signal;
+    const promise = getCaseCounts({ dateFrom, dateTo, country }, true, signal);
+    (promise as PromiseWithCancel<CaseCountEntry[]>).cancel = () => controller.abort();
```
This part cancels the query when unmounted. (See [documentation](https://react-query.tanstack.com/guides/query-cancellation))

ningxie1991 commented on src/components/SwitzerlandEstimatedCasesDivisionModal.tsx:
```diff
@@ -88,23 +92,25 @@ export const SwitzerlandEstimatedCasesDivisionModal = ({
-  }, [dateFrom, dateTo, country, refetch]);
+    if (!isFetching) {
+      refetch();
+    }
```
This `refetch()` is invoked only when `isFetching` is false inside the `useEffect` hook

ningxie1991 committed:
Updated the error handling for getSamples api

- update error handling for getSamples api
- refactored some code to had to change getCaseCount as well

ningxie1991 committed:
Modified the error message display to use Alert

ningxie1991 committed:
Fix formatting

chaoran-chen commented:
By the way, feel free to merge once you have a clean and working version! I'd love to use React Query for future components.

chaoran-chen commented on src/components/SampleTable.tsx:
```diff
@@ -58,17 +61,14 @@ interface Props {
+  const { isLoading, isSuccess, error, isError, data: samples, refetch, isFetching } = useQuery<
+    SampleResultList,
+    Error
+  >('samples', () => {
```
Hi @ningxie1991. I just started to read about React query. Am I understanding it correctly that the query key should always contain all variables, i.e., in this case the pangolin lineage, the mutations string, the match percentage, the country, and the sampling strategy?

https://react-query.tanstack.com/guides/query-keys

ningxie1991 replied:
The query key is an identifier for the particular api request. The parameters for the api request are given in the query function right below:

https://github.com/cevo-public/cov-spectrum-website/blob/f82638d8c3ee4093898dac5bbfd57fd8535ee1ee/src/components/SampleTable.tsx#L72-L78

ningxie1991 replied:
Oh I see what you mean. We should add the dependent variables to the query key as well.

ningxie1991 committed:
Updated error handling using react query

ningxie1991 committed:
Merge branch 'develop' into ui-error-handling

ningxie1991 committed:
Added error handling

ningxie1991 committed:
Updated api error messages

ningxie1991 commented:
> By the way, feel free to merge once you have a clean and working version! I'd love to use React Query for future components.

I merged this pull request.
